Move from Fox to SwiftCheck

gfontenot opened the pull request:
I really like SwiftCheck's syntax more then Fox. And this removes a
metric fuckload of overhead, since it becomes the only dependency.

Right now, this is pointed at `master`, because I need a new release of
SwiftCheck that doesn't conflict with the operators we define.

See https://github.com/typelift/SwiftCheck/issues/100 for more info.

gfontenot committed:
Move from Fox to SwiftCheck

I really like SwiftCheck's syntax more then Fox. And this removes a
metric fuckload of overhead, since it becomes the only dependency.

Right now, this is pointed at `master`, because I need a new release of
SwiftCheck that doesn't conflict with the operators we define.

See https://github.com/typelift/SwiftCheck/issues/100 for more info.

gfontenot committed:
Remove qualified imports

We can just import SwiftCheck since their operators don't conflict with
ours anymore.

gfontenot committed:
Point to actual SwiftCheck release

tonyd256 commented on Tests/Tests/ArraySpec.swift:
```diff
@@ -1,131 +1,83 @@
+        // u <*> (v <*> w) = pure (.) <*> u <*> v <*> w
+        property("composition law") <- forAll { (array: [String]) in
+            let lhs = pure(append) <*> (pure(prepend) <*> array)
+            let rhs = pure(curry(compose)) <*> pure(append)  <*> pure(prepend) <*> array
```
extra space in here

tonyd256 commented:
this looks great

CodaFi commented:
I've released 0.3.1 so you no longer have to tag `master`.

CodaFi commented on Tests/Tests/ArraySpec.swift:
```diff
@@ -1,131 +1,83 @@
+
+        // pure f <*> pure x = pure (f x)
+        property("homomorphism law") <- forAll { (string: String) in
+            let lhs: [String] = pure(append) <*> pure(string)
```
If you want stronger laws, [SwiftCheck can generate functions](https://github.com/typelift/SwiftCheck/blob/master/SwiftCheckTests/ModifierSpec.swift#L45-L59).

gfontenot replied:
Ha, amazing. I had asked about this with Fox, but never got around to working on it. I'll definitely check it out now.

gfontenot replied:
Interestingly, generating functions _seems_ to be giving me a number of intermittent failures. Not sure what that's about.

gfontenot committed:
Generate functions for optional spec

gfontenot committed:
Generate functions for Array spec

gfontenot committed:
Remove helper functions

These aren't needed anymore, since we're generating functions for our
tests.

gfontenot commented:
Updated to generate the functions used in the tests. So everything is comically slow now. It's awesome.

Interestingly, this causes the Applicative composition law to fail for both arrays and optionals, which doesn't seem right. Not sure if @CodaFi can shed some insight or ideas on how to debug/fix?

Failure message:

For arrays:

```
*** Failed! Proposition: composition law
Falsifiable (after 7 tests and 4 shrinks):
[">ÀnÌ¾", "îPêK"]
[String -> String, String -> String, String -> String, String -> String, String -> String, String -> String, String -> String]
[String -> String]
```

For optional:

```
*** Failed! Proposition: composition law
Falsifiable (after 4 tests and 13 shrinks):
Optional("")
Optional(String -> String)
Optional(String -> String)
```

CodaFi commented:
If you want to speed it up, first off I recommend a less complicated type like `Int` or `Int8`.  Second, we've run into this same problem in Swiftz and Swiftx with types like `[String]` because the `Arbitrary` instance is written for Haskell-style representations of `String`s (`[Character]`).  But the majority of the speed issues have to do with the invocation of the function that essentially multiplicatively inflates the array no matter what length it may be each time it encounters `<*>` (trust me, it was once a lot worse.  Swiftz ate all 8GB of my RAM at one point).  The solution I came up with, after optimizing a bunch of stuff in the testing loop, was to just [limit the size of array coming through](https://github.com/typelift/Swiftz/blob/master/SwiftzTests/ListSpec.swift#L79-L94).

gfontenot commented:
Good lord, that is _much_ faster. Thanks for the tip.

The applicative composition laws are still failing for some reason though. I don't _think_ the failures are valid, but it's hard to tell. Is there a good way to dig in deeper?

gfontenot committed:
Use Int instead of String for tests

String is a much more complicated type, so it _really_ slows these tests
down to use it. Int makes the test much much much faster.

CodaFi commented:
I don't have any advice outside of just running the test case with the failing Ints and seeing each step.  I didn't take a look at your implementations, but your laws seem to be equivalent to the ones from Swiftz which do hold.  I would guess something about your Applicative isn't quite right if SwiftCheck is complaining (a law is a law).

If you want more output you can try `forAll {}.verbose`

CodaFi commented:
Yep, just ran your code verbatim in Swiftz's test suite and it passes with flying colors.  Looks like you have a bug on your hands!

gfontenot commented:
![](http://dl.dropboxusercontent.com/u/452364/gifs/fuckit.gif)

I'm merging this with the failing tests.

@CodaFi I'm confused. Do you mean that you think there's a bug in SwiftCheck, or in my implementations of `apply`, or in Swift itself?

CodaFi commented:
You probably have a bug in apply or thereabouts -your laws hold when used with Swiftx/z's definitions.

gfontenot commented:
That... doesn't sound right.

Runes:

``` swift
public func <*> <T, U>(fs: [T -> U], a: [T]) -> [U] {
    return a.apply(fs)
}

public extension Array {
    func apply<U>(fs: [Element -> U]) -> [U] {
        return fs.flatMap { self.map($0) }
    }
}
```

Swiftx:

``` swift
public func <*> <A, B>(fs : [(A -> B)], xs : [A]) -> [B] {
    return fs.flatMap({ xs.map($0) })
}
```

Changing our definition of `<*>` to the Swiftx version (bypassing the instance method) still results in a failure.

CodaFi commented:
Hmph, then it's either the law itself or something we can't see here.  Debugging this is easy enough:

``` swift
        property("composition law") <- forAll { (a: ArrayOf<Int>, fa: ArrayOf<ArrowOf<Int, Int>>, fb: ArrayOf<ArrowOf<Int, Int>>) in
            let x = a.getArray
            let f = fa.getArray.map { $0.getArrow }
            let g = fb.getArray.map { $0.getArrow }

            let lhs = f <*> (g <*> x)
            let rhs = [(curry(compose))] <*> f <*> g <*> x

            return (lhs == rhs).whenFail {
                print("\(lhs) != \(rhs)");
            }
        }.verbose
```

yields

> [0, 0, -1, -1, -1, -1] != [-1, -1, 0, 0, -1, -1]

CodaFi commented:
HA! I've got it!

Your definition of `compose` is backwards.  You probably meant

``` swift
public func compose<A, B, C>(f : B -> C, g : A -> B) -> A -> C {
    return { (a : A) -> C in
        return f(g(a))
    }
}
```

gfontenot commented:
d'oh

I sure did. I wasn't even looking there. My bad, thanks for the help.

CodaFi commented:
Glad to be of service (and to get your tests passing).  CC me into these things or file issues on SwiftCheck if you have any more questions or want to know some fancy tricks sometime.

:sparkles:
